pyro/poutine/trace.py

```diff
@@ -20,103 +19,112 @@ def _warn_if_nan(name, value):
+    def __len__(self):
+        return len(self.nodes)
 
     def __iter__(self):
-        return iter(self._graph)
+        return iter(self.nodes)
 
-    def __len__(self):
-        return len(self._graph)
+    def successors(self, node):
+        return iter(self.edges[node])
 
-    def add_node(self, site_name, *args, **kwargs):
+    def add_node(self, site_name, **kwargs):
         """
         :param string site_name: the name of the site to be added
 
-        Adds a site to the trace.
-
-        Identical to super(Trace, self).add_node,
-        but raises an error when attempting to add a duplicate node
-        instead of silently overwriting.
+        Adds a site with metadata in the kwargs to the trace.
+        Raises an error when attempting to add a duplicate sample node.
         """
         # XXX should do more validation than this
         if kwargs["type"] != "param":
             assert site_name not in self, \
                 "site {} already in trace".format(site_name)
+        self.nodes[site_name] = dict(**kwargs)
+        self.edges[site_name] = {}
 
-        # XXX should copy in case site gets mutated, or dont bother?
-        self._graph.add_node(site_name, *args, **kwargs)
+    def remove_node(self, site_name):
+        """
+        :param string site_name: the name of the site to be removed
+
+        Deletes node site_name and any edges it belongs to.
+        """
+        self.nodes.pop(site_name, None)
+        self.edges.pop(site_name, None)
+
+    def add_edge(self, node_from, node_to, **kwargs):
+        """
+        :param string node_from: the name of the parent site
+        :param string node_to: the name of the child site
+
+        Adds an edge with optional metadata to the trace.
+        Both sites must already be in the trace.
+        """
+        assert node_from in self, \
+            "node_from {} not in trace".format(node_from)
+        assert node_to in self, \
+            "node_to {} not in trace".format(node_to)
+        assert node_to not in self.edges[node_from], \
+            "edge from {} to {} already exists".format(node_from, node_to)
+        self.edges[node_from][node_to] = dict(**kwargs)
 
     def copy(self):
         """
         Makes a shallow copy of self with nodes and edges preserved.
-        Identical to super(Trace, self).copy(), but preserves the type
-        and the self.graph_type attribute
         """
-        trace = Trace()
-        trace._graph = self._graph.copy()
-        trace.graph_type = self.graph_type
-        return trace
+        new_trace = Trace(graph_type=self.graph_type)
+        new_trace.graph.update(self.graph.copy())
+        new_trace.nodes.update(self.nodes.copy())
+        new_trace.edges.update(self.edges.copy())
+        return new_trace
+
+    def iter_edges(self):
+        """
+        Iterates (parent, child) pairs corresponding to edges in the trace.
+        """
+        for node_from in self.edges:
```

Comment:
nit: it's a tiny bit cheaper to iterate over dictionary items rather than keys, because it avoids a hash lookup:
```py
for node_from, edges_from in self.edges.items():
    for node_to in edges_from:
        yield node_from, node_to
```